### PR TITLE
feat: Add Swap Settings SwapProvider

### PR DESCRIPTION
**Pull request by cpcramer**
**What changed? Why?**
## Add `SwapSettings` Component

This PR introduces a new `SwapSettings` component, providing a dropdown interface for users to customize slippage settings during swaps. The component is composed of several subcomponents that manage different aspects of the slippage settings.

Key changes and additions:
- Create `SwapSettings` component with subcomponents:
  - `SwapSettingsSlippageDescription`
  - `SwapSettingsSlippageInput`
  - `SwapSettingsSlippageLayout`
  - `SwapSettingsSlippageLabel`
  - `SwapSettingsSlippageLayoutBottomSheet` for mobile view
- Add `swapSettings` SVG icon
- Update `Swap` component to include and render `SwapSettings`
- Implement click-outside functionality to close the dropdown
- Add full test coverage for all new files and components

Functionality:
- Display settings dropdown when the SVG icon is clicked
- Allow users to toggle between Auto and Custom slippage modes
- Enable custom slippage percentage input
- Provide styling and text override options for various subcomponents

This enhancement gives users more control over their swap settings while maintaining a clean and intuitive interface.
Recording:

https://github.com/user-attachments/assets/68a4a37a-57bf-4c1d-877d-3ccf24149977

**Notes to reviewers**
see design in Slack

Fast follow PRs
- Polish styling

**How has it been tested?**
locally

## Idle states
Screenshot 2024-08-25 at 10 02 05 PM

Screenshot 2024-08-25 at 10 02 44 PM

Screenshot 2024-08-26 at 12 25 43 PM

Screenshot 2024-08-26 at 12 25 50 PM

Screenshot 2024-08-27 at 9 55 25 PM

## Override state
image

## Responsive
image

### CODE REVIEW

File: src/swap/components/Swap.tsx [hunk 1]
```diff
@@ -50,10 +53,14 @@ export function Swap({
         )}
         data-testid="ockSwap_Container"
       >
-        <div className="mb-4">
-          <h3 className={text.title3} data-testid="ockSwap_Title">
+        <div className="mb-4 flex items-center justify-between">
```
Comment (Zizzamia) [inline src/swap/components/Swap.tsx:56]: @cpcramer have you tested how this looks with and without `swapSettings`, so that the CSS is not breaking.

File: src/swap/components/SwapProvider.tsx [hunk 2]
```diff
@@ -58,6 +58,13 @@ export function SwapProvider({
   const { from, to } = useFromTo(address);
   const { sendTransactionAsync } = useSendTransaction(); // Sending the transaction (and approval, if applicable)
```
Comment (Zizzamia) [inline src/swap/components/SwapProvider.tsx:60]: Why are you adding those spaces in line 60 and 64?

Spaces have to be tactical to devide group of code, in this case those Hooks are part of the `// Core Hooks`.

File: src/swap/components/SwapProvider.tsx [hunk 3]
```diff
@@ -231,22 +238,24 @@ export function SwapProvider({
...
     from,
-    loading,
     handleAmountChange,
     handleToggle,
     handleSubmit,
-    lifeCycleStatus,
     isTransactionPending,
+    lifeCycleStatus,
+    loading,
+    maxSlippage,
```
Comment (Zizzamia) [inline src/swap/components/SwapProvider.tsx:256]: Not a fan of adding more things in the Context. As adding more things in the context, shows weakness in the data-flow of information in the architecture.

Let me read what's going on.

File: src/swap/components/SwapProvider.tsx (hunk 2 above)
Comment (cpcramer) [reply src/swap/components/SwapProvider.tsx:60]: Looks like it was an oversight on my end. will fix

File: src/swap/components/Swap.tsx (hunk 1 above)
Comment (cpcramer) [reply src/swap/components/Swap.tsx:56]: Yes, it works as expected!

File: src/swap/components/SwapSettings.tsx [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  useEffect(() => {
+    document.addEventListener('mousedown', handleClickOutsideComponent);
+    return () => {
+      document.removeEventListener('mousedown', handleClickOutsideComponent);
+    };
+  }, [handleClickOutsideComponent]);
+
+  const iconSvg = useIcon({ icon });
+
+  if (!breakpoint) {
```
Comment (Zizzamia) [inline src/swap/components/SwapSettings.tsx:42]: Can you add a line of comment here, I don't understand what are you trying to avoid.

**Comment by Zizzamia** [comment]
Btw feels odd the two inputs have different height. Something we should adjust.
Screenshot 2024-08-28 at 1 35 09 PM

File: src/swap/components/SwapSettingsSlippageInput.tsx [hunk 5]
```diff
@@ -0,0 +1,67 @@
+import { type ChangeEvent, useCallback, useEffect, useState } from 'react';
+import { cn } from '../../styles/theme';
+import type { SwapSettingsSlippageInputReact } from '../types';
+import { useSwapContext } from './SwapProvider';
+
+export function SwapSettingsSlippageInput({
+  className,
+  defaultSlippage = 3,
+  customSlippageEnabled,
```
Comment (Zizzamia) [inline src/swap/components/SwapSettingsSlippageInput.tsx:9]: is this a boolean? if that is the case we should start this prop with `is`

Comment (cpcramer) [reply src/swap/components/SwapSettingsSlippageInput.tsx:9]: yes it's a boolean, will update

File: src/swap/components/SwapSettings.tsx (hunk 4 above)
Comment (cpcramer) [reply src/swap/components/SwapSettings.tsx:42]: yes! This check is following the design in the WalletDropdownBottomSheet. 

I'll test if this is needed, if not then I will remove.

File: src/swap/components/SwapSettingsSlippageInput.tsx [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+}: SwapSettingsSlippageInputReact) {
+  const { maxSlippage, setMaxSlippage } = useSwapContext();
+  const [slippageValue, setSlippageValue] = useState(maxSlippage.toString());
+
+  useEffect(() => {
+    if (customSlippageEnabled) {
```
Comment (Zizzamia) [inline src/swap/components/SwapSettingsSlippageInput.tsx:15]: Add some comment here to help understand what's going on

File: src/swap/components/SwapProvider.tsx (hunk 3 above)
Comment (cpcramer) [reply src/swap/components/SwapProvider.tsx:256]: Agreed that we should be careful about adding info in the Context but I think it makes sense with maxSlippage

File: src/swap/components/SwapSettingsSlippageInput.tsx (hunk 6 above)
Comment (cpcramer) [reply src/swap/components/SwapSettingsSlippageInput.tsx:15]: Will do

File: src/swap/components/SwapSettingsSlippageInput.tsx [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    } else {
+      setSlippageValue(defaultSlippage.toString());
+      setMaxSlippage(defaultSlippage);
+    }
+  }, [customSlippageEnabled, maxSlippage, setMaxSlippage, defaultSlippage]);
+
+  const handleSlippageChange = useCallback(
+    (e: ChangeEvent<HTMLInputElement>) => {
+      const newValue = e.target.value;
+      if (newValue === '' || !Number.isNaN(Number.parseFloat(newValue))) {
```
Comment (Zizzamia) [inline src/swap/components/SwapSettingsSlippageInput.tsx:26]: Add comments here to help understand what you are checking

File: src/swap/components/SwapSettingsSlippageInput.tsx [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        setSlippageValue(newValue);
+        const numericValue = Number.parseFloat(newValue);
+        if (!Number.isNaN(numericValue)) {
```
Comment (Zizzamia) [inline src/swap/components/SwapSettingsSlippageInput.tsx:29]: Same here

File: src/swap/components/SwapSettingsSlippageLayout.tsx [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+import type { SwapSettingsSlippageLayoutReact } from '../types';
+import { SwapSettingsSlippageDescription } from './SwapSettingsSlippageDescription';
+import { SwapSettingsSlippageInput } from './SwapSettingsSlippageInput';
+import { SwapSettingsSlippageTitle } from './SwapSettingsSlippageTitle';
+import { SwapSettingsSlippageToggle } from './SwapSettingsSlippageToggle';
+
+export function SwapSettingsSlippageLayout({
+  children,
+  className,
+  customSlippageEnabled,
```
Comment (Zizzamia) [inline src/swap/components/SwapSettingsSlippageLayout.tsx:13]: boolen start with `is`

File: src/swap/components/SwapSettingsSlippageLayout.tsx [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  onToggleCustomSlippage,
```
Comment (Zizzamia) [inline src/swap/components/SwapSettingsSlippageLayout.tsx:14]: Listeneres are not allowed at component, level, and **all listeners** have to be controled by the `onStatus` lifecycle hook.

File: src/swap/components/SwapSettingsSlippageLayoutBottomSheet.tsx [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+import type { SwapSettingsSlippageLayoutReact } from '../types';
+import { SwapSettingsSlippageDescription } from './SwapSettingsSlippageDescription';
+import { SwapSettingsSlippageInput } from './SwapSettingsSlippageInput';
+import { SwapSettingsSlippageTitle } from './SwapSettingsSlippageTitle';
+import { SwapSettingsSlippageToggle } from './SwapSettingsSlippageToggle';
+
+export function SwapSettingsSlippageLayoutBottomSheet({
+  children,
+  className,
+  customSlippageEnabled,
```
Comment (Zizzamia) [inline src/swap/components/SwapSettingsSlippageLayoutBottomSheet.tsx:13]: boolean start with `is`

File: src/swap/components/SwapSettingsSlippageLayoutBottomSheet.tsx [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+  onToggleCustomSlippage,
```
Comment (Zizzamia) [inline src/swap/components/SwapSettingsSlippageLayoutBottomSheet.tsx:14]: listeners are not allowed in any componet, the only listeners allowed are the one controlled by the lifecyclestatus.

File: src/swap/components/SwapSettingsSlippageLayoutBottomSheet.tsx [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+      ),
+      toggle: childrenArray.find(findComponent(SwapSettingsSlippageToggle)),
+      input: childrenArray.find(findComponent(SwapSettingsSlippageInput)),
+    };
+  }, [children]);
+
+  const enhancedToggle = useMemo(() => {
+    if (isValidElement(toggle)) {
+      return cloneElement(toggle, {
+        customSlippageEnabled,
```
Comment (Zizzamia) [inline src/swap/components/SwapSettingsSlippageLayoutBottomSheet.tsx:31]: This all `slippageEnable` story it's very confusing. I still don't understand why are we making this mess in passing this around.

If the component is there, than we want custom, if it's not there we don't want custom. End of the story. :)

File: src/swap/components/SwapSettingsSlippageTitle.tsx [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+import type { SwapSettingsSlippageTitleReact } from '../types';
+
+export function SwapSettingsSlippageTitle({
+  children,
+  className,
+}: SwapSettingsSlippageTitleReact) {
+  return (
+    <h3
+      className={cn(
+        'mb-2 font-semibold text-base text-gray-950 leading-normal dark:text-gray-50',
```
Comment (Zizzamia) [inline src/swap/components/SwapSettingsSlippageTitle.tsx:11]: `text-gray-950`

wait a second, these all colors will break when custom theming is implemented.

Comment (cpcramer) [reply src/swap/components/SwapSettingsSlippageTitle.tsx:11]: Won't custom theming just override them?

File: src/swap/components/SwapSettingsSlippageLayoutBottomSheet.tsx (hunk 13 above)
Comment (cpcramer) [reply src/swap/components/SwapSettingsSlippageLayoutBottomSheet.tsx:31]: Ok I will clean this up

File: src/swap/components/SwapSettingsSlippageTitle.tsx (hunk 14 above)
Comment (cpcramer) [reply src/swap/components/SwapSettingsSlippageTitle.tsx:11]: Where do you see it breaking? Dark theme seems to be working on playground and in /site

File: src/swap/components/SwapSettings.tsx [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+      setIsOpen(false);
+    }
+  }, []);
+
+  useEffect(() => {
+    document.addEventListener('mousedown', handleClickOutsideComponent);
+    return () => {
+      document.removeEventListener('mousedown', handleClickOutsideComponent);
+    };
+  }, [handleClickOutsideComponent]);
```
Comment (alessey) [inline src/swap/components/SwapSettings.tsx:38]: This seems like something that might be useful to have in a util, i.e. const ref = useOutsideClick(callback)

**Comment by cpcramer** [comment]
> Btw feels odd the two inputs have different height. Something we should adjust. Screenshot 2024-08-28 at 1 35 09 PM

Fixed

File: src/swap/components/SwapSettingsSlippageLayoutBottomSheet.tsx (hunk 13 above)
Comment (cpcramer) [reply src/swap/components/SwapSettingsSlippageLayoutBottomSheet.tsx]: Did a major cleanup. I combined the Input and Toggle sub-components. Was able to remove a ton of logic including customSlippage

File: src/swap/components/SwapSettingsSlippageLayoutBottomSheet.tsx (hunk 12 above)
Comment (cpcramer) [reply src/swap/components/SwapSettingsSlippageLayoutBottomSheet.tsx]: Able to remove these listeners with the refactor

File: src/swap/components/SwapSettingsSlippageLayout.tsx (hunk 10 above)
Comment (cpcramer) [reply src/swap/components/SwapSettingsSlippageLayout.tsx]: removed

File: src/swap/components/SwapSettingsSlippageLayoutBottomSheet.tsx (hunk 12 above)
Comment (cpcramer) [reply src/swap/components/SwapSettingsSlippageLayoutBottomSheet.tsx]: Will start to think Life Cycle Status first

File: src/swap/components/SwapSettingsSlippageLayout.tsx (hunk 10 above)
Comment (cpcramer) [reply src/swap/components/SwapSettingsSlippageLayout.tsx]: will start to think Life Cycle Status first

File: src/swap/components/SwapSettings.tsx (hunk 15 above)
Comment (cpcramer) [reply src/swap/components/SwapSettings.tsx:37]: And it would be nice to standardize our click outside component functionality with the WalletDropdown

File: src/api/buildSwapTransaction.ts [hunk 16]
```diff
@@ -45,6 +45,17 @@ export async function buildSwapTransaction(
     };
   }
 
+  // Adjust slippage for V1 API (aggregator)
+  // V1 expects slippage in tenths of a percent (e.g., 30 = 3%)
+  if (params.useAggregator && params.maxSlippage) {
```
Comment (Zizzamia) [inline src/api/buildSwapTransaction.ts:50]: Let's do this in a different Pr.

File: src/internal/svg/swapSettings.tsx [hunk 17]
```diff
@@ -0,0 +1,21 @@
+import { fill } from '../../styles/theme';
+
+export const swapSettingsSvg = (
```
Comment (Zizzamia) [inline src/internal/svg/swapSettings.tsx:3]: Ok, let's start split this in small PRs. This can go on his own.

File: src/wallet/hooks/useIcon.tsx [hunk 18]
```diff
@@ -8,10 +9,12 @@ export const useIcon = ({ icon }: { icon?: React.ReactNode }) => {
       return null;
     }
     switch (icon) {
-      case 'wallet':
-        return walletSvg;
       case 'fundWallet':
         return fundWalletSvg;
+      case 'swapSettings':
```
Comment (Zizzamia) [inline src/wallet/hooks/useIcon.tsx:14]: Ok, let's start split this in small PRs. This can go on his own.

File: src/swap/components/SwapSettingsSlippageDescription.tsx [hunk 19]
```diff
@@ -0,0 +1,18 @@
+import { cn } from '../../styles/theme';
+import type { SwapSettingsSlippageDescriptionReact } from '../types';
+
+export function SwapSettingsSlippageDescription({
```
Comment (Zizzamia) [inline src/swap/components/SwapSettingsSlippageDescription.tsx:4]: Ok, let's start split this in small PRs. This can go on his own.

File: src/swap/components/SwapSettingsSlippageTitle.tsx [hunk 20]
```diff
@@ -0,0 +1,18 @@
+import { cn } from '../../styles/theme';
+import type { SwapSettingsSlippageTitleReact } from '../types';
+
+export function SwapSettingsSlippageTitle({
```
Comment (Zizzamia) [inline src/swap/components/SwapSettingsSlippageTitle.tsx:4]: Ok, let's start split this in small PRs. This can go on his own.

File: src/swap/components/Swap.tsx [hunk 21]
```diff
@@ -50,10 +52,14 @@ export function Swap({
...
```
Comment (cpcramer) [inline src/swap/components/Swap.tsx:55]: note: be more semantic when possible

File: src/swap/components/SwapSettingsSlippageInput.tsx [hunk 22]
```diff
@@ -33,9 +33,13 @@ export function SwapSettingsSlippageInput({
   // Parses the input and updates slippage if valid
   const handleSlippageChange = useCallback(
     (newSlippage: string) => {
-      const newSlippageNumber = Number.parseFloat(newSlippage);
-      if (!Number.isNaN(newSlippageNumber)) {
-        updateSlippage(newSlippageNumber);
+      if (newSlippage === '') {
+        setSlippage(0);
```
Comment (cpcramer) [inline src/swap/components/SwapSettingsSlippageInput.tsx:37]: Fixes a bug where we weren't able to completely delete the input

File: src/swap/components/SwapProvider.tsx [hunk 23]
```diff
@@ -125,6 +125,11 @@ export function SwapProvider({
     }
   }, [hasHandledSuccess, lifeCycleStatus.statusName]);
 
+  let maxSlippage: string;
+  if (lifeCycleStatus.statusData && 'maxSlippage' in lifeCycleStatus.statusData) {
```
Comment (Zizzamia) [inline src/swap/components/SwapProvider.tsx:129]: I am a bit confused what's going on here.

**Comment by Zizzamia** [comment]
My advice with this PR, as there are a few edge cases we want to keep polish out.

Create one more small PR where you focus only on:
- SwapSettings
- SwapSettingsSlippageInput

But avoid touching SwapProvider.

So we can safelly keep pushing this and even merge it tomorrow. We will do the final thing with SwapProvider after that.

File: src/swap/components/SwapProvider.tsx (hunk 23 above)
Comment (cpcramer) [reply src/swap/components/SwapProvider.tsx]: This is all to avoid Typescript errors. For example, if I just have:

` const maxSlippage = String(lifeCycleStatus.statusData.maxSlippage)`

Then I'll see the following errors: 

```
> -  'lifeCycleStatus.statusData' is possibly 'null'.ts(18047)
> - 'lifeCycleStatus.statusData' is possibly 'null'.ts(18047)
> - Property 'maxSlippage' does not exist on type 'SwapError | { isMissingRequiredField: boolean; } | { amountFrom: string; amountTo: string; tokenFrom?: Token | undefined; tokenTo?: Token | undefined; isMissingRequiredField: boolean; } | { ...; } | { ...; } | { ...; }'.Property 'maxSlippage' does not exist on type 'SwapError'.ts(2339)
```

**Comment by Zizzamia** [comment]
Alright, let's rebase this, and finish up this experience here.

File: src/api/getSwapQuote.ts [hunk 24]
```diff
@@ -16,6 +17,11 @@ import { getAPIParamsForToken } from './utils/getAPIParamsForToken';
 export async function getSwapQuote(
   params: GetSwapQuoteParams,
```
Comment (Zizzamia) [inline src/api/getSwapQuote.ts:18]: Two notes here:
1. show case all params
2. never pass `lifeCycleStatus` inside APIs. API have to stay atomic, and it should be simple input and simple output. Lfecycle is bounded to a component, and the API should be used even without the need to bound something to a component.

Point 2, is a clear point of reflection of System Design interviews or even how to approach the PP2 interview at Coinbase. You want to better polish your thinking on how to design components and APIs, as that is a key skill for any software eng.

Buy this book today https://www.amazon.com/Pragmatic-Programmer-Journeyman-Master/dp/020161622X/ref=sr_1_2?dib=eyJ2IjoiMSJ9.84Xun4zePRzxcYJXHbNXB1wvZkeWhYzsnLiFASxdDssZHty2sHX0n1IsAHY91IBcN34Cmi35fJiQVNMMiFrkKNz3XE4uM_V6mTzQQ-PhrYJA_l1J95Dirj52U0HTs4DKmBxmG0jJPNxAAikRX-jmaXe9b2Ho-w6xebOyzIT4h5N8NhMtGg2mgM20llcF8I5e7GFIt9rwm_T9R1bS_GzZWZ3xiGVss401XLWI_YOUwFI.BheZx0fh6soDsekBonqo9vuAiOIbQS0ibUS2N_HjxfE&dib_tag=se&keywords=The+Pragmatic+Programmer&qid=1725944624&sr=8-2 and read it back to back a couple of times.

You have to polish the way you think about coding so that you can elevate your coding experience to mastery.

File: src/api/buildSwapTransaction.ts [hunk 25]
```diff
@@ -17,6 +18,14 @@ import { getSwapTransaction } from './utils/getSwapTransaction';
 export async function buildSwapTransaction(
   params: BuildSwapTransactionParams,
 ): Promise<BuildSwapTransactionResponse> {
+  const lifeCycleStatus = params.lifeCycleStatus;
+  const maxSlippage =
+    lifeCycleStatus.statusName !== 'error'
+      ? lifeCycleStatus.statusData.maxSlippage
+      : DEFAULT_MAX_SLIPPAGE;
+
+  console.log('buildSwapTransaction maxSlippage ', maxSlippage);
```
Comment (cpcramer) [inline src/api/buildSwapTransaction.ts:27]: remove

File: src/swap/components/SwapProvider.tsx [hunk 26]
```diff
@@ -144,6 +145,10 @@ export function SwapProvider({
       dToken?: Token,
       // biome-ignore lint/complexity/noExcessiveCognitiveComplexity: TODO Refactor this component
     ) => {
+      const maxSlippage =
+        lifeCycleStatus.statusName !== 'error'
+          ? lifeCycleStatus.statusData.maxSlippage
+          : DEFAULT_MAX_SLIPPAGE;
```
[the remark this answers is not in the document]
Comment (cpcramer) [reply src/swap/components/SwapProvider.tsx:151]: good call! Updated

File: src/swap/components/SwapProvider.tsx [hunk 27]
```diff
@@ -47,7 +48,7 @@ export function SwapProvider({
   // Feature flags
   const { useAggregator } = experimental;
   const [maxSlippage, _setMaxSlippage] = useState(
-    experimental.maxSlippage || 3,
+    experimental.maxSlippage || DEFAULT_MAX_SLIPPAGE,
```
[the remark this answers is not in the document]
Comment (cpcramer) [reply src/swap/components/SwapProvider.tsx:51]: updated the variable to `initialMaxSlippage`

File: src/swap/components/SwapProvider.tsx [hunk 28]
```diff
@@ -68,6 +69,12 @@ export function SwapProvider({
   // Refreshes balances and inputs post-swap
   const resetInputs = useResetInputs({ from, to });
 
+  const getMaxSlippage = useCallback(() => {
+    if (lifeCycleStatus.statusData && lifeCycleStatus.statusName !== 'error') {
+      return lifeCycleStatus.statusData.maxSlippage;
+    }
+    return experimental.maxSlippage || DEFAULT_MAX_SLIPPAGE;
+  }, [lifeCycleStatus, experimental.maxSlippage]);
```
Comment (Zizzamia) [inline src/swap/components/SwapProvider.tsx:77]: I feel like there is something logically wrong here.... I am going to read this more carefully.

Comment (Zizzamia) [reply src/swap/components/SwapProvider.tsx:77]: I feel like, the fact that `statusData` it's not the same when an `error` occurs, creates this weird odd logic, which will have more side effects.

Like, if an error occurs, we lose whatever was the last `maxSliappage` a user might have customized.

Plus, adding those weird if/else, creates a precedent where for other options we need to play this odd if/else

Comment (Zizzamia) [reply src/swap/components/SwapProvider.tsx:77]: Let me open a PR to help write down my thinking.

Comment (Zizzamia) [reply src/swap/components/SwapProvider.tsx:77]: https://github.com/coinbase/onchainkit/pull/1234

File: src/swap/components/SwapProvider.tsx (hunk 27 above)
[the remark this answers is not in the document]
Comment (Zizzamia) [reply src/swap/components/SwapProvider.tsx:51]: @cpcramer why do we still need this code?

should just `experimental.maxSlippage || DEFAULT_MAX_SLIPPAGE` go to like 62?

in the initial value for LifeCycleStatus?

**Review by Zizzamia — APPROVED** [review]
Amazing, and I left only one note that could be done as fast-follow.

**Review by alessey — APPROVED** [review]
